libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1119,6 +1119,28 @@ void SMTEncoder::arrayPushPopAssign(Expression const& _expr, smtutil::Expression
 	}
 	else if (auto const* indexAccess = dynamic_cast<IndexAccess const*>(&_expr))
 		arrayIndexAssignment(*indexAccess, _array);
+	else if (auto const* funCall = dynamic_cast<FunctionCall const*>(&_expr))
+	{
+		FunctionType const& funType = dynamic_cast<FunctionType const&>(*funCall->expression().annotation().type);
+		if (funType.kind() == FunctionType::Kind::ArrayPush)
+		{
+			auto memberAccess = dynamic_cast<MemberAccess const*>(&funCall->expression());
+			solAssert(memberAccess, "");
+			auto symbArray = dynamic_pointer_cast<smt::SymbolicArrayVariable>(m_context.expression(memberAccess->expression()));
+			solAssert(symbArray, "");
+
+			auto oldLength = symbArray->length();
+			auto store = smtutil::Expression::store(
+				symbArray->elements(),
+				symbArray->length() - 1,
```

Comment:
Shouldn't this be just `length()`?

Reply:
One annoying part of this implementation is that in `array.push().push()` the inner `array.push()` will run first, so it's length will already be adjusted. Here we only assign the last item to the result of the right `push()`.

Reply:
The code you're referring to is above in `void SMTEncoder::arrayPush(FunctionCall const& _funCall)`. The code is very similar, so I'm still trying to find a way to merge those pieces of code and avoid repetition.

Reply:
This looks to me like a special case that should be solved in a more general way. What if you have something like
```
function f(... x) internal returns (...) { return x.push(); }
function g() internal {
  f(...).push();
}
```

Reply:
Yea also things like
```
uint[][] a;
function f(uint x) internal view returns (uint[] storage) { returns a[x]; }
function g(uint x) public { f(x).push(); }
```
wouldn't be handled by that.
The problem with these 2 cases is that storage pointers erase knowledge anyway